<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Perfect match aligner</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for matchAlign {MIPSAlign}"><tr><td>matchAlign {MIPSAlign}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Perfect match aligner</h2>

<h3>Description</h3>

<p>Generates counts on perfect sequence matches
</p>


<h3>Usage</h3>

<pre>
matchAlign(seqData = list.files(pattern = "*.fastq.gz$"), orfLib, numCores = 1)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>seqData</code></td>
<td>
<p>The list of Fastq files you want to align</p>
</td></tr>
<tr valign="top"><td><code>orfLib</code></td>
<td>
<p>The reference Fasta alignment file</p>
</td></tr>
<tr valign="top"><td><code>numCores</code></td>
<td>
<p>The number of cores to run on</p>
</td></tr>
</table>


<h3>Value</h3>

<p>A list of aligned and unaligned counts dataframes, respectively
</p>


<h3>Examples</h3>

<pre>
x &lt;- matchAlign(orfLib = fastaFileName,numCores = 6)
</pre>

<hr /><div style="text-align: center;">[Package <em>MIPSAlign</em> version 0.0.0.9000 <a href="00Index.html">Index</a>]</div>
</body></html>
